added redme (#2)
A1eksD/baby-shop: README.md

# E-Commerce Project For Baby Tools
A small project with basic CRUD operations. In this application you can create baby-product listings and sort them in the admin area. The app was built with Django, which is based on Python and lets you spin up the project quickly.

## Table of Contents

1. [Prerequisites](#prerequisites)
1. [Hints](#hints)
1. [Download Repository](#download-repository)
1. [Install Docker](#install-docker)
1. [Create Dockerfile](#create-dockerfile)
1. [Build Docker Image](#build-docker-image)
1. [Start Docker Container](#start-docker-container)
1. [How to connect](#how-to-connect)
1. [Photos](#photos)
1. [Useful commands](#useful-commands)

## Prerequisites

- Python 3.9
- Django 4.0.2
- Venv
- Docker
- git

---

## Hints
This section will cover some hot tips when trying to interacting with this repository:

- Settings & Configuration for Django can be found in `babyshop_app/babyshop/settings.py`
- Routing information, such as available routes can be found from any `urls.py` file in `babyshop_app` and corresponding subdirectories

---

> ## Download Repository

Reviewer: Rename it to `Quickstart` and do not forget to update the Table of contents sections.

# E-Commerce Project For Baby Tools

> A samll project with basic CRUD operatiosn.

Reviewer 2: Add more details about the project:

- What is the purpose of this web application?
- What are the used technologies?
- What features does it offer?

For caching typos in English words, you can use this VS Code extension that highlights misspelled words. [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)

```suggestion
A small project with basic CRUD operations.
```

## Table of Contents

1. [Prerequisites](#prerequisites)
1. [Hints](#hints)
1. [Photos](#photos)

> 1. [Usage](#usage)

Reviewer 2: The Usage Link does not work.

  - [Install Venv and Project](#install-venv-and-project)
  - [Run Docker](#run-docker)
  - [Edit container](#edit-container)
  - [How to connect](#how-to-connect)

> ### Prerequisites

Reviewer 2: Prerequisites is not a subsection of Table of Contents.
```suggestion
## Prerequisites
```

- Python 3.9
- Django 4.0.2

> - Venv

Reviewer 2: Docker should also be listed here.
```suggestion
## Prerequisites

- Python 3.9
- Django 4.0.2
- Venv
- Docker
```

---

### Hints
py
This section will cover some hot tips when trying to interacting with this repository:

- Settings & Configuration for Django can be found in `babyshop_app/babyshop/settings.py`
- Routing: Routing information, such as available routes can be found from any `urls.py` file in `babyshop_app` and corresponding subdirectories

> 

Reviewer 2: Also, not a subsection of Table of Contents. 
Remove typos `py`.
Remove `Routing:`. Makes unnecessary noise on reading.

```suggestion
## Hints

This section will cover some hot tips when trying to interacting with this repository:

- Settings & Configuration for Django can be found in `babyshop_app/babyshop/settings.py`
- Routing information, such as available routes can be found from any `urls.py` file in `babyshop_app` and corresponding subdirectories

```

---

### Photos

##### Home Page with login

<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080815407.jpg"></img>
##### Home Page with filter
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080840305.jpg"></img>
##### Product Detail Page
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080934541.jpg"></img>

##### Home Page with no login
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080953570.jpg"></img>

##### Register Page

<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323081016022.jpg"></img>

##### Login Page

> <img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323081044867.jpg"></img>

Reviewer 2: Also, not a subsection of Table of Contents. 
and correct the subsections of Photos.

```suggestion
## Photos

### Home Page with login

<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080815407.jpg"></img>
### Home Page with filter
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080840305.jpg"></img>
### Product Detail Page
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080934541.jpg"></img>

### Home Page with no login
<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323080953570.jpg"></img>

### Register Page

<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323081016022.jpg"></img>

### Login Page

<img alt="" src="https://github.com/MET-DEV/Django-E-Commerce/blob/master/project_images/capture_20220323081044867.jpg"></img>
```

---

> 

Reviewer 2: Suggested section structure that follows Table of Contents section :

```suggestion
## Quickstart

1. Download Repository
1. Create Python Virtual Environment
1. Activate Python Virtual Environment
1. Install Requirements
1. Initialize Database
1. Start the application
```

> ## Install Venv and Project

Reviewer 2: This section presumes you already have the project on your machine, so I suggest you first start by explaining how to first clone the project.

> [!Note]

> > local

Reviewer 2: I presume you mean here that this Setup should be performed on your local machine. But it should not matter. The instructions should work regardless of Environment.
```suggestion
```

> #### Open in your Terminal/Powershell and got to your project folder and insert the folowing prompt:

Reviewer 2: The previous sections should already guide you to be in a Terminal and at the project location.

```suggestion
1. Create Python Virtual Environment
```

```powershell
python -m venv venv
```

> #### Then start your venv.

Reviewer 2: I would recommend avoiding use of `Then`. It implicitly exists as this are steps to follow

```suggestion
1. Start the Python Virtual Environment.
```

Windows:
```powershell
.\venv\Scripts\activate
```
Mac / Linux:
```powershell
source ./venv/bin/activate
```
> [!Note]

> > Check the path for yourself.

Reviewer 2: To wage, what path to check?
```suggestion
> Replace the path to the Python Virtual Environment to match yours.
```

#### Copy now the project withs HTTPS or SSH in your Folder.
```powershell
git clone <the key>

> ```

Reviewer 2: This section should be place much earlier in the guide and again avoid the `now` as this are steps to follow, words like `then`, `now` they follow implicitly. And you can use your Repository Link here in the clone command.

```suggestion
1. Download the project.

```powershell
git clone https://github.com/A1eksD/baby-shop
```
```

> #### Install now the requirements:

Reviewer 2: ```suggestion
1. Install the requirements:
```

```powershell
pip install -r </path/to>/requirements.txt
```

> #### With this prompt you can start the applicaltion and check it out:

Reviewer 2: ```suggestion
1. Start the applicaltion and check it out:
```

```powershell
python manage.py runserver
```
> [!Note]
> You need to be in the folder with your manage.py for this prompt.

---

> 

Reviewer 2: Suggested section that follows Quickstart section:

```suggestion
## Usage

### Using Docker

1. Install Docker
1. Create `Dockerfile`
1. Build Docker Image
1. Start Docker Container

```

## Create Docker file

#### if Docker is not yet installed.
```powershell
sudo apt update
sudo apt upgrade
sudo apt install docker.io
```

> #### Create a file named `Dockerfile` at the same level as the `.gitignore`.

Reviewer 2: ```suggestion
1. Install Docker

```powershell
sudo apt update
sudo apt upgrade
sudo apt install docker.io
```

1. Create Docker file at the same level as the `.gitignore`.
```

- The file should look like this:

```powershell
# 1. Base image

> FROM python:3.10-alpine

Reviewer 2: Here you are pointing to a python3.10 Image, but Prerequisites point to python3.9 version.
```suggestion
FROM python:3.9-alpine
```

# 2. Useful Python defaults
ENV PYTHONDONTWRITEBYTECODE=1 \
    PYTHONUNBUFFERED=1

# 3. Working directory (project root)
WORKDIR /app

# 4. Copy only requirements first – keeps the cache valid
COPY requirements.txt .

RUN pip install --no-cache-dir -r requirements.txt

# 5. Copy the rest of the application code

> COPY . $WORKDIR

Reviewer 2: Use **{}-Notation** 
```suggestion
COPY . ${WORKDIR}
```

# 6. Expose port 8082
EXPOSE ${APP_PORT}

# 7. Declare a persistent volume
#    /app/data - save in host-sytem
VOLUME /app/data
#    /app/data/media - place to save Uploads/SQLite files
ENV MEDIA_ROOT=/app/data/media

# 8. On container start:
#    • Create and apply migrations
#    • Create / update the Django super-user
#    • Launch the Django development server on port 8082
#      (Use Gunicorn/Uvicorn instead of runserver for production)
CMD ["sh", "-c", "python babyshop_app/manage.py makemigrations --noinput && \
                  python babyshop_app/manage.py migrate --noinput && \
                  python babyshop_app/manage.py createsuperuser --noinput \
                        --username ${DJANGO_SUPERUSER_USERNAME} \
                        --email ${DJANGO_SUPERUSER_EMAIL} || true && \
                  python babyshop_app/manage.py runserver 0.0.0.0:${APP_PORT}"]

```

> - Also check your `.dockerignore` if the file fit your requirements.

Reviewer 2: This does not fit the current documentation. It presumes that the Project already has a `.dockerignore` file, then why a guide to create a `Dockerfile`?

A. You remove the `Create Dockerfile` step entirely, and you link in the README to the already existing `Dockerfile` and keep this line. I prefer this one.

Or you also create a step on how to create a `.dockerignore` file. But if you do so, then every time your Dockerfile is changed, you need to update the README.

--- 

## Run Docker
- If you want to save files (optional) 
```powershell
docker volume create <folder_name>

> ```

Reviewer 2: ```suggestion
## Run Docker

1. Create Docker Volume (optional) 
```powershell
docker volume create <volume_name>
```
```

- Run container on server
```powershell
docker run -p ${APP_PORT}:${APP_PORT} \
  -e DJANGO_SUPERUSER_USERNAME=${ADMIN_NAME} \
  -e DJANGO_SUPERUSER_PASSWORD=${ADMIN_PASSWORD} \
  -e DJANGO_SUPERUSER_EMAIL=${ADMIN_EMAIL} \
  babyshop
```

> 

Reviewer 2: You are missing the step to create a Docker Image.
```suggestion
1. Creat Docker Image
1. Run Docker container
```powershell
docker run -p ${APP_PORT}:${APP_PORT} \
  -e DJANGO_SUPERUSER_USERNAME=${ADMIN_NAME} \
  -e DJANGO_SUPERUSER_PASSWORD=${ADMIN_PASSWORD} \
  -e DJANGO_SUPERUSER_EMAIL=${ADMIN_EMAIL} \
  babyshop
```

```

---

## Edit container

#### create volume
```powershell
docker volume create babyshop_db
```
#### create
```powershell
docker volume create <volume-name>
```

#### shows all volumes
```powershell
docker volume ls
```

#### delete one or more
```powershell
docker volume rm <volume-name>
```

> ---

Reviewer 2: Can be placed after the Usage Section as list with useful commands.

## How to connect

> #### Go to your browser and search for `<server-ip>:<port>`.

Reviewer 2: This section should be placed at the end of Usage section.

# E-Commerce Project For Baby Tools
A small project with basic CRUD operations. In this application you can create baby-product listings and sort them in the admin area. The app was built with Django, which is based on Python and lets you spin up the project quickly.

## Table of Contents

1. [Prerequisites](#prerequisites)
1. [Hints](#hints)
1. [Quickstart](#quickstart)
  - [Download Repository](#download-repository)
1. [Usage Docker](#usage-docker)
  - [Install Docker](#install-docker)
  - [Create Dockerfile](#create-dockerfile)
  - [Build Docker Image](#build-docker-image)
  - [Start Docker Image](#start-docker-image)
  - [How to connect](#how-to-connect)
1. [Photos](#photos)

> 1. [Useful commands](#useful-commands)

Reviewer 2: Keep it simple.

```suggestion
1. [Prerequisites](#prerequisites)
1. [Hints](#hints)
1. [Quickstart](#quickstart)
1. [Usage Docker](#usage-docker)
1. [Photos](#photos)
1. [Useful commands](#useful-commands)
```

## Prerequisites

- Python 3.9
- Django 4.0.2
- Venv
- Docker
- git

---

## Hints
This section will cover some hot tips when trying to interacting with this repository:

- Settings & Configuration for Django can be found in `babyshop_app/babyshop/settings.py`
- Routing information, such as available routes can be found from any `urls.py` file in `babyshop_app` and corresponding subdirectories

---

## Quickstart

1. Download Repository
```powershell
git clone https://github.com/A1eksD/baby-shop
```

2. Create Python Venv
```powershell
python -m venv venv

> ```

Reviewer 2: Between step 1 and 2 you should also advise navigating to the project directory:

  2. Navigate to project Directory
  ```powershell
  cd  baby-shop
  ```

3. Activate Python Venv
Windows:
```powershell
.\venv\Scripts\activate
```
Mac / Linux:
```powershell
source ./venv/bin/activate
```
> [!Note]

> > Replace the path to the Python Virtual Environment ro match yours.

Reviewer 2: ```suggestion
> Replace the path to the Python Virtual Environment to match yours.
```

4. Install Requirements
```powershell
pip install -r </path/to>/requirements.txt
```

5. Install Database
```powershell
python manage.py makemigrations
python manage.py migrate
```
> [!Note]
> You need to be in the folder with your manage.py for this prompt.

6. Start the application
```powershell
python manage.py runserver
```
> [!Note]
> You need to be in the folder with your manage.py for this prompt.

> 

Reviewer 2: You are repeating the same note twice. You should generally try to avoid repeating things. Just place the Note at the end of the section, advising that running `manage.py` Commands should always be run from the same folder where `manage.py` file exists.

CMD ["sh", "-c", "python babyshop_app/manage.py makemigrations --noinput && \
                  python babyshop_app/manage.py migrate --noinput && \
                  python babyshop_app/manage.py createsuperuser --noinput \
                        --username ${DJANGO_SUPERUSER_USERNAME} \
                        --email ${DJANGO_SUPERUSER_EMAIL} || true && \
                  python babyshop_app/manage.py runserver 0.0.0.0:${APP_PORT}"]

```

> [!Note]
> Also check whether you need/want to have a `.dockerignore` file. Place it at the same level as the `Dockerfile`. If you're not sure whether you need one, you can use this template: https://github.com/A1eksD/baby-shop/blob/main/.dockerignore

--- 

### Start Docker Image
1. Create Docker Volume (optional) 
```powershell
docker volume create <volume_name>
```

2. Create Docker Image
```powershell
docker build -t babyshop:latest --build-arg APP_PORT=${APP_PORT} .
```

3. Run container on server
```powershell
docker run -p ${APP_PORT}:${APP_PORT} \
  -e DJANGO_SUPERUSER_USERNAME=${ADMIN_NAME} \
  -e DJANGO_SUPERUSER_PASSWORD=${ADMIN_PASSWORD} \
  -e DJANGO_SUPERUSER_EMAIL=${ADMIN_EMAIL} \
  babyshop
```

> ---

Reviewer 2: ```suggestios

1. Build Docker Image

```powershell
docker build --build-arg APP_PORT=${CONTAINER_PORT} -t babyshop .
```

1. Start Docker Container

```powershell
docker run -p ${HOST_PORT}:${CONTAINER_PORT} \
  -e DJANGO_SUPERUSER_USERNAME=${ADMIN_NAME} \
  -e DJANGO_SUPERUSER_PASSWORD=${ADMIN_PASSWORD} \
  -e DJANGO_SUPERUSER_EMAIL=${ADMIN_EMAIL} \
  babyshop
```

1.  Open the application in your browser `<server-ip>:<port>`

---

> ### How to connect
> 
> Search in your browser for `<server-ip>:<port>`.

Reviewer 2: ```suggestion
```
